### DISCUSSION TITLE
Compiler overview and structure

### DISCUSSION THREAD

**Discussion by normrubin** [normrubin/normrubin.github.io#70] [class related]
questions and comments

**Comment by aymane-eljerari** [comment]
What are things to consider when trying to quantify the performance of a single compiler optimization. What are the implications and differences between the two following experiments:

1. Use a program compiled using -O1 as a baseline. Then only disable a single flag from the list of -O1 flags. This is to evaluate the performance degredation that comes from disabling only one flag from -O1.
2. Use a program compiled with -O0 as a basline. Then enable a single flag from the list of -O1 flags to evaluate the performance speed up that comes from that optimization alone.

How much "independence" do compiler optimizations have? Are their performance mostly additive? If not, which I believe is most likely the case, how can we effectively benchmark the performance of a single optimization flag?

**Reply by normrubin** [reply]
optimizations are not additive- There are often complex interactions for example dead code elimination might be more effective after inlining, There are also phase ordering issues. Some optimizations rely on others being applied first, and different orderings of optimizations can lead to varying performance outcomes

baselines: using -01 as as base and removing one optimization means many other optimizations are already applied - so the effect of removing one might be masked by the others

using -o0 as the the base means that few optimizations are applied and some might be more effective when used in a group

it is quite complex to tell if a single optimization helps or hinders-  You often find a wack-a-mole result, Adding an optimization makes some programs run faster but makes others run slower,  so deciding if an optimization helps, may depend of which specific programs you care about
